File: Android_Zone_Lib/src/com/zone/lib/utils/activity_fragment_ui/handler/HandlerUiUtil.java
package com.zone.lib.utils.activity_fragment_ui.handler;

import android.os.Handler;
import android.os.Looper;

/**
 * @author MaTianyu
 * @date 2015-03-12
 */
public class HandlerUiUtil {
    public static final Handler HANDLER = new Handler(Looper.getMainLooper());

    public static void post(Runnable runnable) {
        HANDLER.post(runnable);
    }

    public static void postDelayed(Runnable runnable, long delayMillis) {
        HANDLER.postDelayed(runnable, delayMillis);
    }

    public static void removeCallbacks(Runnable runnable) {
        HANDLER.removeCallbacks(runnable);
    }
}

File: Android_Zone_Lib/src/com/zone/lib/utils/activity_fragment_ui/handler/HandlerUiUtil.kt
package com.zone.lib.utils.activity_fragment_ui.handler

import android.os.Handler
import android.os.Looper
import java.lang.ref.WeakReference

/**
 * @author MaTianyu
 * @date 2015-03-12
 */
object HandlerUiUtil {
    val HANDLER = Handler(Looper.getMainLooper())

    fun post(runnable: Runnable) {
        HANDLER.post(runnable)
    }

    fun postDelayed(runnable: Runnable, delayMillis: Long) {
        HANDLER.postDelayed(runnable, delayMillis)
    }

    fun removeCallbacks(runnable: Runnable) {
        HANDLER.removeCallbacks(runnable)
    }

    fun postSafeExist(runnable: Runnable, obj: Any?) {
        postDelayedExist(runnable, 0L, obj)
    }

    fun postDelayedExist(runnable: Runnable, delayMillis: Long, objIsExist: Any?) {
        val weakObj = WeakReference(objIsExist)
        val runableProxy = Runnable {
            weakObj.get()?.let {
                runnable.run()
            }
        }
        if (delayMillis == 0L) HANDLER.post(runableProxy)
        else HANDLER.postDelayed(runableProxy, delayMillis)

    }

    fun postSafe(runnable: Runnable, isSafe: () -> Boolean) {
        postSafeDelayed(runnable, 0L, isSafe)
    }

    fun postSafeDelayed(runnable: Runnable, delayMillis: Long, isSafe: () -> Boolean) {
        val runableProxy = Runnable {
            if (isSafe()) {
                runnable.run()
            }
        }
        if (delayMillis == 0L) HANDLER.post(runableProxy)
        else HANDLER.postDelayed(runableProxy, delayMillis)
    }
}
